---
  title: "Antimicrobial treatment GWAS using Pyseer"
author: "Gerry Tonkin-Hill"
date: "`r Sys.Date()`"
output: 
  html_document:
  fig_width: 12
fig_height: 8
editor_options: 
  chunk_output_type: console
---
  
```{r setup, include = FALSE}
knitr::opts_chunk$set(fig.width=20, fig.height=12,
                      echo=TRUE, warning=FALSE, message=FALSE,
                      tidy=TRUE)
options(stringsAsFactors = FALSE)
```

##Load Libraries

```{r}
library(data.table)
library(tidyverse)
library(broom)
library(ggrepel)
library(parallel)
library(pbapply)
library(lubridate)
```

##Load data

Initially we make use of the unitig count matrix described in the `antimicrobial_treatment_pairwise_gwas.Rmd` file.


```{r}
conservative_carriage_estimates <- fread("./data/carriage_episode_estimates.csv", data.table = FALSE) %>% as_tibble()
manifests_qc <- fread("./data/manifest_w_QC.csv", data.table = FALSE) %>% as_tibble()
swab_summary <- fread("./data/epi/Pneumo-DEEP_swab_summary.csv", data.table = FALSE) %>% as_tibble
abx_summary <- fread("./data/epi/Pneumo-DEEP_illness_abx_summary.csv", data.table = FALSE) %>% as_tibble

swab_summary$specdate <- lubridate::as_date(swab_summary$specdate)
swab_summary$ID <- paste(swab_summary$codenum, swab_summary$category, sep = "-")

#map samples with treatment status
abx_summary$ID <- paste(abx_summary$codenum, "Infant", sep="-")
abx_summary$date_start <- lubridate::as_date(abx_summary$date_start)

swab_summary$treatment <- map_chr(1:nrow(swab_summary), function(i){
  sampleID <- swab_summary$ID[[i]]
  treatments <- abx_summary[abx_summary$ID==sampleID,]
  treatments <- treatments[!is.na(treatments$drug1),]
  diff <- swab_summary$specdate[[i]]-treatments$date_start
  if (any((diff>0) & (diff<=28))){
    return("TREATED")
  } else if (any((diff>0) & (diff<60))) {
    return("AMBIGUOUS")
  } else {
    return("NOT_TREATED")
  }
})

swab_summary <- swab_summary[!duplicated(swab_summary$specnum),]
swab_summary <- swab_summary %>% filter(specnum %in% manifests_qc$`DONOR ID (required for EGA)`)
swab_summary$lane <- manifests_qc$`Lane Name`[match(swab_summary$specnum, manifests_qc$`DONOR ID (required for EGA)`)]
swab_summary <- swab_summary %>% filter(category=="Infant")

swab_summary$is_treated <- swab_summary$treatment=="TREATED"
write.csv(swab_summary, file="./data/swab_summary_pyseer.csv", row.names = FALSE, quote = FALSE)
```

We now create a Kinship matrix to control population structure when using Pyseer.

```{r, eval=FALSE}
#kinship
unitigs_matrix <- data.matrix(all_unitigs[,2:ncol(all_unitigs)])
kinship <- crossprod(unitigs_matrix>0)
write.csv(kinship, file = "combined_counts_filtered_min10_simple_kinship.csv", row.names = TRUE, quote = FALSE)
```

##Running Pyseer

We can also compare presence/absence rather than counts. In this case we are better to make use of a linear mixed model with a kinship matrix. We make use of the pyseer python package for this step of the analysis.


First we need to prepare the input files for Pyseer.

```{r, eval=FALSE}
kin <- fread('combined_counts_filtered_min10_simple_kinship.csv') %>% as_tibble()
colnames(kin) <- gsub("_kmers", "", colnames(kin))
kin$V1 <- gsub("_kmers", "", kin$V1)
write.table(kin, file = './pyseer_run/combined_counts_filtered_min10_simple_kinship.tab', 
            sep = '\t', quote = FALSE, row.names = FALSE, col.names = TRUE)
filt_unitigs <- fread('combined_counts_filtered_min10_simple.csv') %>% as_tibble()
for (i in 2:ncol(filt_unitigs)){
  filt_unitigs[,i] <- 1*(filt_unitigs[,i]>0)  
}
colnames(filt_unitigs) <- gsub("_kmers", "", colnames(filt_unitigs))

fwrite(filt_unitigs, file = './pyseer_run/combined_counts_filtered_min10_simple.tab', 
       sep = '\t', quote = FALSE, row.names = FALSE, col.names = TRUE)

manifests_qc <- fread("./data/manifest_w_QC.csv") %>% as_tibble()
swab_summary_pyseer <- fread(file="./data/swab_summary_pyseer.csv") %>% as_tibble()

swab_summary_pyseer$is_treated <- 1*swab_summary_pyseer$is_treated
swab_summary_pyseer$sample <- swab_summary_pyseer$lane
write.table(swab_summary_pyseer[,c('lane', 'is_treated')], file='./pyseer_run/phenotypes.tab', 
            sep = "\t", quote=FALSE, row.names = FALSE, col.names = TRUE)

write.table(swab_summary_pyseer[,c('lane', '')], file='./pyseer_run/covariates.tab', 
            sep = "\t", quote=FALSE, row.names = FALSE, col.names = TRUE)
```

We can now run pyseer using it's Linear Mixed Model (LMM) pipeline.

```{bash, eval=FALSE}
nohup pyseer --lmm --phenotypes phenotypes.tab --pres combined_counts_filtered_min10_simple.tab --similarity combined_counts_filtered_min10_simple_kinship.tab --output-patterns unitig_patterns_05.txt --min-af 0.05 --max-af 0.95 --cpu 60 > treatment_unitigs_filt05.txt > nohup_pyseer_05.txt &

python ~/software/pyseer/scripts/count_patterns.py unitig_patterns.txt

cat <(head -1 treatment_unitigs.txt) <(awk '$4<8.45e-07 {print $0}' treatment_unitigs.txt) > top_unitigs.txt
```

To control for the number of test we investigate the number of distinct unitig patterns as recommended in the Pyseer documentation.

Patterns:       5915610
Threshold:      8.45E-09 (EFD=0.05)
Threshold:      1.69E-08 (EFD=0.1)
Threshold:      1.69e-07 (EFD=1)
Threshold:      8.45e-07 (EFD=5)

We now take those Unitigs that pass the adjusted p-value cut-off and extract them from the original set generated using Bifrost.

```{python, eval=FALSE}
import pyfastx

hits = {}
with open('top_unitigs.txt', 'r') as infile:
  header = next(infile)
  for line in infile:
    line=line.strip().split()
    hits[line[0]] = line

with open('top_unitigs_pyseer.txt', 'w') as outfile:
  outfile.write(header)
  for h,s in pyfastx.Fasta("../unitig_bifrost.fasta", build_index=False):
    if h in hits:
      l = hits[h]
      l[0] = s
      outfile.write("\t".join(l)+"\n")
```

Similar to the pairwise analysis we use Pyseer to annotate the significant unitigs

```{bash, eval=FALSE}
annotate_hits_pyseer top_unitigs_pyseer.txt pyseer_references.txt annotated_unitigs.txt
```

##Investigate results

We addition to running Pyseer we manually annotated the significant unitigs using the BLAST NCBI database.

```{r}
top_pyseer_results_w_anno <- fread('./data/annotated_unitigs.txt', 
                                   col.names = c('variant','af','filter-pvalue','lrt-pvalue',
                                                 'beta','beta-std-err','variant_h2','annotation')) %>% 
  as_tibble() %>% 
  arrange(`lrt-pvalue`)

top_pyseer_results_w_anno$gene <- map_chr(str_split(top_pyseer_results_w_anno$annotation, ';'), ~ ifelse(length(.x)>1,.x[[2]], .x))


manual_annotations <- c('Streptococcus_pneumoniae_Xen35_GCF_002843545_1_01594'='FtsX-like permease',
                        'Streptococcus_pneumoniae_NT_ST4133_v1_01924'='IS1202 transposase',
                        '7038_5#21_00300'='FtsX-like permease')

top_pyseer_results_w_anno$manual_annotation <- manual_annotations[top_pyseer_results_w_anno$gene]
top_pyseer_results_w_anno[is.na(top_pyseer_results_w_anno)] <- top_pyseer_results_w_anno$gene[is.na(top_pyseer_results_w_anno$manual_annotation)]

write.csv(top_pyseer_results_w_anno, './data/pyseer_top_hit_with_annotations.csv', quote = FALSE, row.names = FALSE)
knitr::kable(top_pyseer_results_w_anno)
```

```{r}
gene_hits <- top_pyseer_results_w_anno %>% 
  group_by(manual_annotation) %>%
  summarise(
      avg_beta = mean(beta),
      min_beta = min(beta),
      max_beta = max(beta),
      maxp = max(-log10(`lrt-pvalue`)),
      hits = n()
  ) %>% arrange(-maxp)

pthresholds <- tibble(p.value=-log10(c(1.69e-08,1.69e-07,8.45e-07)),
       EFD=factor(c(0.1,1,5)))

ggplot(gene_hits, aes(x=avg_beta, y=maxp, size=hits, label=manual_annotation)) +
  geom_vline(xintercept = 0, col='black', alpha=0.3, size=0.7) +
  geom_hline(aes(yintercept=p.value, col=EFD), data = pthresholds, size=0.7, linetype="dashed") +
  geom_point() +
  scale_color_manual(values = c('#de2d26','#fc9272','#fee0d2')) +
  geom_text_repel(aes(size=14), show.legend = FALSE, colour='black', box.padding = 1) +
  scale_x_continuous(limits = c(-0.5,0.5)) +
  scale_y_continuous(limits = c(6,8.5), breaks = c(6,6.5,7,7.5,8,8.5)) +
  scale_size_continuous("Number of unitigs", range=c(3,8), limits=c(0,150), breaks = c(25,50,75,100)) +
  theme_clean(base_size = 20) +
  theme(plot.background = element_blank(),
        legend.background = element_blank()) +
  xlab("Average effect size") +
  ylab("Maximum -log10(p-value)")

ggsave("./figures/pyseer_treatment_gwas_effect_vs_p.png", width = 12, height = 7)
ggsave("./figures/pyseer_treatment_gwas_effect_vs_p.pdf", width = 12, height = 7,
       useDingbats=FALSE)
```