---
title: "Diversity analysis"
output: html_document
author: "Gerry Tonkin-Hill"
date: "`r Sys.Date()`"
output: 
  html_document:
    fig_width: 12
    fig_height: 8
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=8,
                      echo=TRUE, warning=FALSE, message=FALSE,
                      tidy=TRUE)
options(stringsAsFactors = FALSE)
```

### Load data and libraries

```{r, warning=FALSE}
library(tidyverse)
library(data.table)
library(lme4)
library(ggthemes)

swab_summary <- fread("./data/epi/Pneumo-DEEP_swab_summary.csv", data.table = FALSE) %>% as_tibble
abx_summary <- fread("./data/epi/Pneumo-DEEP_illness_abx_summary.csv", data.table = FALSE) %>% as_tibble
merged_serotype_calls <- fread("./data/merged_serotype_calls.csv", data.table = FALSE) %>% as_tibble()
conservative_carriage_estimates <- fread("./data/carriage_episode_estimates.csv", data.table = FALSE) %>% as_tibble()
household_data <- fread("./data/epi/Pneumo-DEEP_infant_house_metadata.csv") %>% as_tibble()
```

## Lineage level

Collect data

```{r}
swab_summary$specdate <- lubridate::as_date(swab_summary$specdate)
swab_summary$ID <- paste(swab_summary$codenum, swab_summary$category, sep = "-")

#map samples with treatment status
abx_summary$ID <- paste(abx_summary$codenum, "Infant", sep="-")
abx_summary$date_start <- lubridate::as_date(abx_summary$date_start)

#add treatment information
swab_summary$treatment <- map_chr(1:nrow(swab_summary), function(i){
  sampleID <- swab_summary$ID[[i]]
  treatments <- abx_summary[abx_summary$ID==sampleID,]
  treatments <- treatments[!is.na(treatments$drug1),]
  diff <- swab_summary$specdate[[i]]-treatments$date_start
  if (any((diff>0) & (diff<=28))){
    return("TREATED")
  } else if (any((diff>0) & (diff<60))) {
    return("AMBIGUOUS")
  } else {
    return("NOT_TREATED")
  }
})


swab_summary <- swab_summary[!duplicated(swab_summary$specnum),]
merged_serotype_calls$treatment <- swab_summary$treatment[match(merged_serotype_calls$specnum, swab_summary$specnum)]
merged_serotype_calls$codenum <- swab_summary$codenum[match(merged_serotype_calls$specnum, swab_summary$specnum)]
merged_serotype_calls$host <- swab_summary$ID[match(merged_serotype_calls$specnum, swab_summary$specnum)]
merged_serotype_calls$specdate <- swab_summary$specdate[match(merged_serotype_calls$specnum, swab_summary$specnum)]

#count lineages per sample and add household variables
lineage_counts <- merged_serotype_calls %>% 
  group_by(specnum, host, treatment,codenum) %>%
  summarise(
    nlineages = n(), .groups = 'keep'
  )

lineage_counts <- merge(lineage_counts, household_data, by.x='codenum', by.y = 'codenum', all.x = TRUE) %>% 
  as_tibble()
lineage_counts$specdate <- swab_summary$specdate[match(lineage_counts$specnum, swab_summary$specnum)]
lineage_counts$age <- lineage_counts$specdate - lubridate::as_date(
  household_data$infant_dob[match(lineage_counts$codenum, household_data$codenum)])
```

The main force of infection appears to be other children <5yr in the house.

```{r}
#look at children
lineage_counts_infant <- lineage_counts %>% filter(grepl('Infant', host))

model <- glmer(nlineages ~ age + infant_gender + house_fire + num_child_less5y + 
                 mother_smoke + num_child_5_14y + num_adult_15_59y + (1|codenum), 
               family = poisson, data = lineage_counts_infant)

summary(model)

# can also consider a model that does not account for household which gives similar results
model <- glm(nlineages ~ age + gender + house_fire + num_child_less5y + 
                 mother_smoke + num_child_5_14y + num_adult_15_59y, 
               family = poisson, data = lineage_counts_infant)

summary(model)
```

We can also make a 'clone' plot to demonstrate how the frequency of lineages changes over time within a single infant.

```{r}
temp <- merged_serotype_calls %>%
  group_by(host) %>%
  summarise(
    num=length(unique(specdate)),
    mmoi=mean(table(specdate)),
    diff=max(diff(unique(specdate)))
    
  ) %>%
  arrange(-num)
temp

example_infant <- "ARI-0905-Infant"


sero_subset <- merged_serotype_calls %>% 
  filter(host == example_infant) %>%
  filter(!is.na(consensus_percentage))
  

swab_subset <- swab_summary %>% 
  filter(paste(codenum, category, sep='-')==example_infant) %>%
  filter(!specnum %in% sero_subset$specnum) %>%
  group_split(specnum) %>%
  map_dfr(~{
    seros <- c(.x$serotype1, .x$serotype2, .x$serotype3, .x$serotype4)
    seros <- seros[!is.na(seros)]
    if (length(seros)<1) seros <- c(NA)
    return(tibble(
      host = paste(.x$codenum, .x$category, sep='-'),
      specnum = .x$specnum,
      specdate = .x$specdate,
      serotype = seros
    ))
  })

swab_subset$frequency <- 0.02
swab_subset$frequency[is.na(swab_subset$serotype)] <- NA

sero_subset <- sero_subset[,c('host','specnum','specdate','serotype','consensus_percentage')]
colnames(sero_subset) <- colnames(swab_subset)

swab_subset$latex <- TRUE
sero_subset$latex <- FALSE
swab_subset <- rbind(swab_subset, sero_subset)

swab_subset2 <- swab_subset %>%
  group_split(specnum, specdate, latex) %>%
  map_dfr(~{
    .x$frequency <- .x$frequency/sum(.x$frequency)
    return(.x)
  })

dob <- as_date(household_data$infant_dob[match(gsub('-Infant','',example_infant), household_data$codenum)])

tpn <- unique(swab_subset2$specdate)
tp <- as.numeric(tpn-dob)
for (i in 2:length(tp)){
  if (tp[i]-tp[i-1]<20){
    tp[i] <- tp[i-1] + 20
  }
}

names(tp) <- tpn  
swab_subset2$date <- tp[as.character(swab_subset2$specdate)]
swab_subset2$serotype[is.na(swab_subset2$serotype)] <- 'unknown'
swab_subset2$method <- ifelse(swab_subset2$latex, 'latex sweep', 'sequencing')

ggplot(swab_subset2, aes(x=date, y=frequency, fill=serotype, alpha=method)) +
  geom_col(width = 15) +
  theme_clean(base_size = 20) +
  scale_alpha_manual(values = c(0.5,1)) +
  theme(plot.background = element_blank(),
        legend.background = element_blank()) +
  scale_fill_manual(values = c('#a6cee3','#1f78b4','#33a02c','#fb9a99','#b2df8a',
                               '#e31a1c','#fdbf6f','#ff7f00','#cab2d6','#6a3d9a',
                               '#ffed6f','#b15928','#80cdc1')) +
  ylab('frequency in host') +
  xlab('age (days)')

ggsave('./figures/serotype_frequency_proportion_barplot_example.png', width = 12, height = 7)
ggsave('./figures/serotype_frequency_proportion_barplot_example.pdf', width = 12, height = 7)
```


## Locus level

Here, we concentrate on those swabs involving only a single lineage. We also split by carriage event.

```{r}
only_minority_single_gpsc_dedup <- fread("./data/only_minority_single_gpsc.tab") %>% 
  as_tibble() %>%
  filter(is.na(region)) %>%
  filter(support>11) %>% 
  filter(depth<1000) %>%
  filter(alt_forward>1) %>%
  filter(alt_reverse>1)

only_minority_single_gpsc_dedup$carriage_episode <- conservative_carriage_estimates$carriage_episode[match(
  only_minority_single_gpsc_dedup$specnum, conservative_carriage_estimates$specnum)]
only_minority_single_gpsc_dedup$specdate <- swab_summary$specdate[match(only_minority_single_gpsc_dedup$specnum, swab_summary$specnum)]
```

We now consider if there is an association between carriage duration and number of minority variants.

```{r}
variant_counts <- only_minority_single_gpsc_dedup %>% 
  group_by(specnum, carriage_episode, specdate, host) %>%
  summarise(
    nvar = n(), .groups = 'keep'
  )

variant_counts$days_from_first <- map2_dbl(variant_counts$carriage_episode, variant_counts$specdate, ~{
  .y-min(variant_counts$specdate[variant_counts$carriage_episode==.x])
})

variant_counts_post <- variant_counts %>% filter(days_from_first>0)
variant_counts_post <- variant_counts_post %>% filter(nvar<100)
variant_counts_post$days_from_first_norm <- scale(variant_counts_post$days_from_first, center = FALSE)

model <- glmer(nvar ~ days_from_first_norm + (1|carriage_episode), 
      data=variant_counts_post, family=poisson)

summary(model)

plot(variant_counts_post$days_from_first, variant_counts_post$nvar)
```

Look at other covariates

```{r}
variant_counts <- only_minority_single_gpsc_dedup %>% 
  group_by(specnum, carriage_episode, specdate, host) %>%
  summarise(
    nvar = n(),
    .groups = 'keep'
  )

variant_counts$days_from_first <- map2_dbl(variant_counts$carriage_episode, variant_counts$specdate, ~{
  .y-min(variant_counts$specdate[variant_counts$carriage_episode==.x])
})

variant_counts_post <- variant_counts %>% filter(nvar<100)
variant_counts_post$days_from_first_norm <- scale(variant_counts_post$days_from_first, center = FALSE)

variant_counts_post$mother <- grepl('Mother', variant_counts_post$host)
variant_counts_post$codenum <- gsub('-[MI].*','', variant_counts_post$host)
variant_counts_post$treated <- merged_serotype_calls$treatment[match(variant_counts_post$specnum, merged_serotype_calls$specnum)]
variant_counts_post$treated <- variant_counts_post$treated=="TREATED"
variant_counts_post$gender <- household_data$infant_gender[match(variant_counts_post$codenum, household_data$codenum)]
variant_counts_post$age <- scale(variant_counts_post$specdate - as_date(household_data$infant_dob[match(
  variant_counts_post$codenum, household_data$codenum)]), center = FALSE)


model <- glmer(nvar ~ days_from_first_norm + mother + treated + age + gender + (1|carriage_episode), 
      data=variant_counts_post, family=poisson)

summary(model)
```

Interestingly, being treated with antibiotics is significantly associated with a decrease in diversity as might be expected given what we observed in the unitig analysis. Somewhat unexpectedly, adults (mothers) had a significantly higher level of diversity to that of infants.

```{r}
pdf <- variant_counts_post 

pdf$type <- ifelse(pdf$mother, 'Mother', 'Infant')
ggplot(pdf, aes(x=type, y=nvar)) +
  ggbeeswarm::geom_quasirandom(width = 0.2) +
  geom_boxplot(outlier.colour = NA, alpha=0) +
  theme_clean(base_size = 20) +
  theme(plot.background = element_blank(),
        legend.background = element_blank()) +
  ylab('number of within-host variants') +
  xlab('')

ggsave('figures/diversity_boxplot_mother_v_infant.png', width = 10, height = 7)
ggsave('figures/diversity_boxplot_mother_v_infant.pdf', width = 10, height = 7)
```

```{r}
pdf <- variant_counts_post
pdf$type <- ifelse(pdf$treated, 'Treated', 'Not Treated')
ggplot(pdf, aes(x=type, y=nvar)) +
  ggbeeswarm::geom_quasirandom(width = 0.2) +
  geom_boxplot(outlier.colour = NA, alpha=0) +
  theme_clean(base_size = 20) +
  theme(plot.background = element_blank(),
        legend.background = element_blank()) +
  ylab('number of within-host variants') +
  xlab('')  

ggsave('figures/diversity_boxplot_treatment.png', width = 10, height = 7)
ggsave('figures/diversity_boxplot_treatment.pdf', width = 10, height = 7)
```

Example within host variant frequency plot

```{r}

only_minority_single_gpsc_dedup %>% group_by(carriage_episode) %>%
  summarise(num=length(unique(specdate))) %>%
  arrange(-num)

example_carriage <- 'ARI-0212_Infant_47_1'

pdf <- only_minority_single_gpsc_dedup %>%
   filter(carriage_episode==example_carriage)
pdf$var <- paste(pdf$chrom, pdf$POS)

ggplot(pdf, aes(x=specdate, y=AF, group=var, col=var)) +
  geom_point() +
  geom_line() +
  theme_clean(base_size = 20) +
  scale_alpha_manual(values = c(0.5,1)) +
  theme(plot.background = element_blank(),
        legend.background = element_blank(),
        legend.position = 'none') +
  xlab('sample date') +
  ylab('allele frequency')

ggsave('./figures/allele_frequency_proportion_barplot_example.png', width = 12, height = 7)
ggsave('./figures/allele_frequency_proportion_barplot_example.pdf', width = 12, height = 7)
```
