---
title: "Resistance and virulence calls"
author: "Gerry Tonkin-Hill"
date: "`r Sys.Date()`"
output: 
  html_document:
    fig_width: 12
    fig_height: 8
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(fig.width=20, fig.height=12,
                      echo=TRUE, warning=FALSE, message=FALSE,
                      tidy=TRUE)
options(stringsAsFactors = FALSE)
```
  
To determine resistance elements we ran the CDC's pneumococcal resistance pipeline on each the deconvoluted samples generated using the mGEMS pipeline. This approach was verified using the laboratory mixture data of Knight et al., 2019 and by comparing to samples where single colony picks had also been sequenced in the study of Chewapreecha et al., 2014. This indicated that a small number of misclassified reads could lead to resistance being called in the incorrect lineage within a mixed samples. As a result we only report resistance calls at the sample level.

We also ran the Ariba resistance calling pipeline using the comprehensive antibiotic resistance database (CARD) (McArthur et al., 2013). We supplemented this database with a custom set of known *S. pneumoniae* resistance elements which have previously been used with Ariba to call resistance in *S. pneumoniae* (Lo et al., 2019).


### CDC resistance pipeline

Similar to the mGEMs pipeline we only report the commands used to run the resistance pipelines for reproducibility purposes and the scripts are not set up to be run from this Rmarkdown file. All the databases used in this section are available on the Zenodo repository that accompanies this manuscript.

We initially generate a list of fastq files that are then passed to the CDC pipeline bash script.

```{bash, eval=FALSE}
for r1 in [0-9]*_1.fastq.gz
do
prefix=$(basename $f _1.fastq.gz)
r2="${r1/_1.fastq.gz/_2.fastq.gz}"
echo ${r1} >> fastq_pair_list.txt
echo ${r2} >> fastq_pair_list.txt
done

Sanger-JanOw_SPN-wrapr.sh -f fastq_pair_list.txt -r /fullpath/to/SPN_Reference_DB-1.0.0/ -o ./CDC_pipeline_output/
```


### Ariba resistance pipeline

First we ran the Ariba pipeline using the amended CARD database.

```{bash, eval=FALSE}
for r1 in [0-9]*_1.fastq.gz
do
prefix=$(basename $f _1.fastq.gz)
r2="${r1/_1.fastq.gz/_2.fastq.gz}"
ariba run ariba_db_Steph_28Mar17_card $r1 $r2 "./resistance/${prefix}_ariba"
done
```

To investigate the associations with non-CPS virulence factors we also ran Ariba using the Virulence Factor database (Chen et al., 2016). 

```{bash, eval=FALSE}
for r1 in [0-9]*_1.fastq.gz
do
prefix=$(basename $f _1.fastq.gz)
r2="${r1/_1.fastq.gz/_2.fastq.gz}"
ariba run vfdb_full_db $r1 $r2 "./virulence/${prefix}_ariba"
done
```


## Results

Initially load some libraries we will need.

```{r}
library(tidyverse)
library(data.table)
library(lubridate)
library(lme4)
library(broom)
library(patchwork)
library(ggridges)
library(ggthemes)
```

Load the results of running the CDC pipeline and some additional metadata.

```{r}
CDC_resistance <- fread("./data/TABLE_SPN_CDC_v1.0.3_output_20200527_Typing_Results.txt", 
                        data.table = FALSE, sep="\t", fill=TRUE) %>% as_tibble()
manifests_qc <- fread("./data/manifest_w_QC.csv", data.table = FALSE) %>% as_tibble()
ena_accessesions <- fread("./data/pneumoDEEP_accessions.csv", data.table = FALSE, 
                          header = FALSE, col.names = c("lane", "ena_acc")) %>% as_tibble()
swab_summary <- fread("./data/epi/Pneumo-DEEP_swab_summary.csv", data.table = FALSE) %>% as_tibble
abx_summary <- fread("./data/epi/Pneumo-DEEP_illness_abx_summary.csv", data.table = FALSE) %>% as_tibble
merged_serotype_calls <- fread("./data/merged_serotype_calls.csv", data.table = FALSE) %>% as_tibble()
```

Similar to the GWAS analysis we assign each swab as being pre or post treatment if the swab was taken within 4 weeks of an antibiotic treatment event. 

```{r}
swab_summary$specdate <- lubridate::as_date(swab_summary$specdate)
swab_summary$ID <- paste(swab_summary$codenum, swab_summary$category, sep = "-")

#map samples with treatment status
abx_summary$ID <- paste(abx_summary$codenum, "Infant", sep="-")
abx_summary$date_start <- lubridate::as_date(abx_summary$date_start)

#add treatment information
swab_summary$treatment <- map_chr(1:nrow(swab_summary), function(i){
  sampleID <- swab_summary$ID[[i]]
  treatments <- abx_summary[abx_summary$ID==sampleID,]
  treatments <- treatments[!is.na(treatments$drug1),]
  diff <- swab_summary$specdate[[i]]-treatments$date_start
  if (any((diff>0) & (diff<=28))){
    return("TREATED")
  } else if (any((diff>0) & (diff<60))) {
    return("AMBIGUOUS")
  } else {
    return("NOT_TREATED")
  }
})

#add diagnosis information
swab_summary$diagnosis <- map(1:nrow(swab_summary), function(i){
  sampleID <- swab_summary$ID[[i]]
  treatments <- abx_summary[abx_summary$ID==sampleID,]
  treatments <- treatments[!is.na(treatments$drug1),]
  treatments <- treatments[abs(swab_summary$specdate[[i]]-treatments$date_start)<28,, drop=FALSE]
  if (nrow(treatments)>0){
    return(unique(treatments$diagnosis_clinical))
  } else {
    return(NA)
  }
})

swab_summary <- swab_summary[!duplicated(swab_summary$specnum),]

merged_serotype_calls$treatment <- swab_summary$treatment[match(merged_serotype_calls$specnum, swab_summary$specnum)]
merged_serotype_calls$diagnosis <- swab_summary$diagnosis[match(merged_serotype_calls$specnum, swab_summary$specnum)]
merged_serotype_calls$host <- swab_summary$ID[match(merged_serotype_calls$specnum, swab_summary$specnum)]
```

Here we clean up the results a bit and merge the different data files.

```{r}
CDC_resistance$lane <- gsub("_[0-9]*$", "", CDC_resistance$Sample)
CDC_resistance$GPSC <- as.numeric(gsub(".*_", "", CDC_resistance$Sample))
temp <- ena_accessesions$lane[match(CDC_resistance$lane, ena_accessesions$ena_acc)]
CDC_resistance$lane[!is.na(temp)] <- temp[!is.na(temp)]
CDC_resistance$specnum <- manifests_qc$`DONOR ID (required for EGA)`[match(CDC_resistance$lane, manifests_qc$`Lane Name`)]

CDC_resistance$consensus_percentage <- merged_serotype_calls$consensus_percentage[
  match(paste(CDC_resistance$specnum,
              CDC_resistance$GPSC),
        paste(merged_serotype_calls$specnum,
              merged_serotype_calls$GPSC))]

CDC_resistance$treated <- merged_serotype_calls$treatment[
  match(paste(CDC_resistance$specnum, CDC_resistance$GPSC),
        paste(merged_serotype_calls$specnum,
              merged_serotype_calls$GPSC))]

CDC_resistance$host <- merged_serotype_calls$host[
  match(paste(CDC_resistance$specnum, CDC_resistance$GPSC),
        paste(merged_serotype_calls$specnum,
              merged_serotype_calls$GPSC))]

CDC_resistance <- CDC_resistance %>% 
  filter(paste(specnum, GPSC) %in% 
           paste(merged_serotype_calls$specnum, merged_serotype_calls$GPSC))

a <- colnames(CDC_resistance)[grepl("SIR", colnames(CDC_resistance))]
a <- a[map_lgl(a, ~ {
  temp <- unique(unlist(CDC_resistance[,colnames(CDC_resistance)==.x]))
  temp <- temp[!is.na(temp)]
  temp <- temp[!temp %in% c('Flag','I','NF')]
  return(length(temp)>1)
})]
```


## Verification of Pipeline

We can now verify the pipeline using the known mixtures of Knight et al., 2019 and the single colony picks of Chewapreecha et al., 2014.

```{r}
known_single_CDC <- fread("./data/known_single_TABLE_SPN_CDC_pipeline_Typing_Results.txt", data.table = FALSE)
known_mixed_CDC <- fread("./data/known_mixed_TABLE_SPN_CDC_pipeline_Typing_Results.txt", data.table = FALSE)

known_single_CDC$serotype <- map(known_single_CDC$Sample, ~{
  unlist(str_split(gsub(".*Serotype_", "", .x), "_"))
})

known_mixed_CDC <- known_mixed_CDC %>% filter(grepl(".*Serotype_mixture.*", Sample))
known_mixed_CDC$serotypes <- map(known_mixed_CDC$Sample, ~{
  s <- unlist(str_split(gsub(".*mixture_", "", .x), "_"))
  return(s[-length(s)])
})

known_mixed_CDC$mixed_sample <- map_chr(known_mixed_CDC$Sample, ~ gsub("_[0-9]*$", "", .x))

keep <- unlist(map(c('R'), ~ paste(a, .x)))
known_cdc_comparison_res <- map_dfr(unique(known_mixed_CDC$mixed_sample), function(s){
  # if (grepl("23F", s)) return(NULL)
  mix <- known_mixed_CDC %>% filter(mixed_sample==s)
  if (length(unique(unlist(mix$serotypes)))>Inf) return(tibble())
  res <- unique(unlist(map(a, ~ paste(.x, mix[,.x]))))
  res_single <- unique(unlist(map(unique(unlist(mix$serotypes)), ~{
    # if(.x=="23F") return(sero23F)
    single <- known_single_CDC %>% filter(serotype==.x)
    if (nrow(single)<1) return(NULL)
    return(paste(a, single[,a]))
  })))
  res <- res[res %in% keep]
  res_single <- res_single[res_single %in% keep]
  tibble(
    sample=s,
    TP=sum(res %in% res_single),
    FP=sum(!res %in% res_single),
    FN=sum(!res_single %in% res)
  )
})

plotdf <- tibble(
  variable=factor(c('TP','FP','FN'), levels=c('TP','FP','FN')),
  value=c(sum(known_cdc_comparison_res$TP),
          sum(known_cdc_comparison_res$FP),
          sum(known_cdc_comparison_res$FN))
)
precision <- 1-sum(known_cdc_comparison_res$FP)/sum(known_cdc_comparison_res$FP+known_cdc_comparison_res$TP)
recall <- sum(known_cdc_comparison_res$TP)/sum(known_cdc_comparison_res$FN+known_cdc_comparison_res$TP)

rename <- factor(setNames(c('True Positive', 'False Positive', 'False Negative'), c('TP','FP','FN')),
                 levels = c('True Positive', 'False Positive', 'False Negative'))
plotdf$variable <- rename[plotdf$variable]

ggplot(plotdf, aes(x=variable, y=value)) + geom_col() +
  theme_clean(base_size = 16) +
  theme(plot.background = element_blank(),
        legend.background = element_blank()) +
  ylab("") +
  xlab("")+
  annotate("text", x=3, y=25, label= sprintf("precision: %.3f", precision), size=5) + 
  annotate("text", x=3.09, y=23, label= sprintf("recall: %.3f", recall), size=5) +
  theme(legend.position = "none")

ggsave("./figures/CDC_calls_known_mix_precision_recall.png", width = 10, height = 7)
ggsave("./figures/CDC_calls_known_mix_precision_recall.pdf", width = 10, height = 7)
```

Looking at the results at the level of the deconvoluted lineages in the known mixtures we can see that while the pipeline is generally able to correctly distinguish resistance into the correct lineage there are occasional cases where the resistance call gets misassigned within a mixed sample. For this reason we only call resistance features at the sample level. We don't consider the mixture involving 10 strains as we do not observe this level of multiplicity of infection in the dataset and the mGEMs pipeline is not able to detect the very low frequency strains in this sample.

```{r}
known_cdc_comparison_res <- map_dfr(unique(known_mixed_CDC$mixed_sample), function(sample){
  mix <- known_mixed_CDC %>% filter(mixed_sample==sample)
  if (length(unique(unlist(mix$serotypes)))>8) return(tibble())
  res <- unique(unlist(map(a, ~ paste(.x, mix[,.x]))))
  res_single <- unique(unlist(map(unique(unlist(mix$serotypes)), ~{
    single <- known_single_CDC %>% filter(serotype==.x)
    if (nrow(single)<1) return(NULL)
    return(paste(a, single[,a]))
  })))
  res <- res[res %in% keep]
  res_single <- res_single[res_single %in% keep]
  tibble(
    sample=sample,
    TP=sum(res %in% res_single),
    FP=sum(!res %in% res_single),
    FN=sum(!res_single %in% res)
  )
})

plotdf <- tibble(
  variable=factor(c('TP','FP','FN'), levels=c('TP','FP','FN')),
  value=c(sum(known_cdc_comparison_res$TP),
          sum(known_cdc_comparison_res$FP),
          sum(known_cdc_comparison_res$FN))
)
precision <- 1-sum(known_cdc_comparison_res$FP)/sum(known_cdc_comparison_res$FP+known_cdc_comparison_res$TP)
recall <- sum(known_cdc_comparison_res$TP)/sum(known_cdc_comparison_res$FN+known_cdc_comparison_res$TP)

plotdf$variable <- rename[plotdf$variable]

ggplot(plotdf, aes(x=variable, y=value)) + geom_col() +
  theme_clean(base_size = 16) +
  theme(plot.background = element_blank(),
        legend.background = element_blank()) +
  ylab("") +
  xlab("")+
  annotate("text", x=3, y=15, label= sprintf("precision: %.3f", precision), size=5) + 
  annotate("text", x=3.09, y=14, label= sprintf("recall: %.3f", recall), size=5) +
  theme(legend.position = "none")

ggsave("./figures/CDC_calls_known_mix_collapsed_at_sample_precision_recall.png", width = 10, height = 7)
ggsave("./figures/CDC_calls_known_mix_collapsed_at_sample_precision_recall.pdf", width = 10, height = 7)
```

We can now compare this approach to the resistance calls found in the whole genome assemblies of the single colony picks of Chewapreecha et al., 2014. 

```{r}
single_colony_cdc <- fread("./data/single_colony_TABLE_SPN_CDC_pipeline_Typing_Results.txt", data.table = FALSE)
single_deep_pairs <- fread("./data/single_deep_pairs.csv", data.table = FALSE) %>% as_tibble

keep <- unlist(map(c('S','I','R'), ~ paste(a, .x)))

single_vs_deep_cdc <- map_dfr(1:nrow(single_deep_pairs), function(i){
  specnum <- single_deep_pairs$specnum[[i]]
  single <- single_deep_pairs$single[[i]]
  deep <- single_deep_pairs$deep[[i]]
  
  single_cdc <- single_colony_cdc %>% filter(Sample==single)
  deep_cdc <- CDC_resistance %>% filter(lane==deep)
  
  if ((nrow(single_cdc)<1) | (nrow(deep_cdc)<1)) return(tibble())
  
  single_cdc <- unique(unlist(map(a, ~ paste(.x, unlist(single_cdc[,.x])))))
  deep_cdc <- unique(unlist(map(a, ~ paste(.x, unlist(deep_cdc[,.x])))))
  
  single_cdc <- single_cdc[single_cdc %in% keep]
  deep_cdc <- deep_cdc[deep_cdc %in% keep]
  
  tibble(
    specnum=specnum,
    single_S_only = sum((!single_cdc %in% deep_cdc) & grepl(".* S", single_cdc)),
    single_I_only = sum((!single_cdc %in% deep_cdc) & grepl(".* I", single_cdc)),
    single_R_only = sum((!single_cdc %in% deep_cdc) & grepl(".* R", single_cdc)),
    deep_S_only = sum((!deep_cdc %in% single_cdc) & grepl(".* S", deep_cdc)),
    deep_I_only = sum((!deep_cdc %in% single_cdc) & grepl(".* I", deep_cdc)),
    deep_R_only = sum((!deep_cdc %in% single_cdc) & grepl(".* R", deep_cdc)),
    both_R = sum((single_cdc %in% deep_cdc ) & grepl(".* R", single_cdc)),
    both = sum(single_cdc %in% deep_cdc),
    TP=sum(single_cdc %in% deep_cdc),
    FP=sum(!deep_cdc %in% single_cdc),
    FN=sum(!single_cdc %in% deep_cdc)
  )
})

plotdf <- tibble(
  variable=factor(c('TP','FN'), levels=c('TP','FN')),
  value=c(sum(single_vs_deep_cdc$both_R),
          sum(single_vs_deep_cdc$single_R_only))
)

plotdf$variable <- rename[plotdf$variable]

ggplot(plotdf, aes(x=variable, y=value, fill=variable)) + geom_col() +
  theme_clean(base_size = 16) +
  theme(plot.background = element_blank(),
        legend.background = element_blank()) +
  ylab("") +
  xlab("") +
  annotate("text", x=2, y=3300, label= sprintf("recall: %0.3f", sum(single_vs_deep_cdc$both_R)/(sum(single_vs_deep_cdc$both_R)+sum(single_vs_deep_cdc$single_R_only) )), size=5) +
  scale_fill_manual(values = c('#756bb1','#bcbddc'), guide=FALSE)

ggsave("./figures/single_vs_deep_cdc_calls_precision_recall.pdf", width = 10, height = 7)
ggsave("./figures/single_vs_deep_cdc_calls_precision_recall.png", width = 10, height = 7)
```

We can also compare the total resistance calls via single colony picks and deep sequencing to investigate the added resolution it provides as a surveillance technique.

```{r}
single_colony_cdc <- fread("./data/single_colony_TABLE_SPN_CDC_pipeline_Typing_Results.txt", data.table = FALSE)
single_deep_pairs <- fread("./data/single_deep_pairs.csv", data.table = FALSE) %>% as_tibble

res_single_deep_counts <- map2_dfr(single_deep_pairs$specnum, single_deep_pairs$single, ~{
  if (!.x %in% CDC_resistance$specnum) {return(tibble())}
  deep <- 1*(colSums(CDC_resistance[CDC_resistance$specnum==.x,a]=='R')>0)
  single <- 1*(colSums(single_colony_cdc[single_colony_cdc$Sample==.y,a]=='R')>0)
  tibble(
    type=rep(c('single','deep'), each=length(a)),
    specnum=rep(.x, 2*length(a)),
    antibiotic=c(names(single), names(deep)),
    res=c(single, deep)
  )
}) %>% group_by(type, antibiotic) %>%
  summarise(
    res_count=sum(res),
    total_count=n()
  )

res_single_deep_counts$antibiotic <- gsub("_SIR", "", gsub("WGS_", "", res_single_deep_counts$antibiotic))

diff_order <- res_single_deep_counts %>% 
  group_by(antibiotic) %>%
  summarise(
    diff_prop=max(abs(diff(res_count)))/sum(res_count[type=='single'])
  ) %>%
  arrange(-diff_prop)
res_single_deep_counts$antibiotic <- factor(res_single_deep_counts$antibiotic, levels = diff_order$antibiotic)

single_sum <- sum(res_single_deep_counts$res_count[res_single_deep_counts$type=='single'])
deep_sum <- sum(res_single_deep_counts$res_count[res_single_deep_counts$type=='deep'])
deep_sum-single_sum
(deep_sum-single_sum)/single_sum

mean(diff_order$diff_prop)

res_single_deep_counts$type[res_single_deep_counts$type=='single'] <- 'single colony'
res_single_deep_counts$type[res_single_deep_counts$type=='deep'] <- 'plate sweep'

ggplot(res_single_deep_counts, aes(x=antibiotic, y=res_count, fill=type)) + 
  geom_col(position = 'dodge') +
  theme_clean(base_size = 16) +
  theme(plot.background = element_blank(),
        legend.background = element_blank()) +
  scale_fill_manual(values=c('#ca0020', '#0571b0')) +
  scale_x_discrete(guide = guide_axis(n.dodge = 2)) +
  scale_y_sqrt(breaks=c(1,10,100,500,1000)) +
  xlab('antibiotic category') +
  ylab("number of samples with resistance determinants")

ggsave("./figures/res_counts_deep_vs_single.png", width = 12, height = 7)
ggsave("./figures/res_counts_deep_vs_single.pdf", width = 12, height = 7)
```

We can also plot the subset which comprise only a single lineage. Here, we expect very similar calls and any deviation would indicate possible false positives in a particular antibiotic class.

```{r}
single_infections <- unique((fread("./data/only_minority_single_gpsc_dedup.tab") %>% as_tibble())$specnum)
temp_pairs <- single_deep_pairs[single_deep_pairs$specnum %in% single_infections,]

temp_res_single_deep_counts <- map2_dfr(temp_pairs$specnum, temp_pairs$single, ~{
  if (!.x %in% CDC_resistance$specnum) {return(tibble())}
  deep <- 1*(colSums(CDC_resistance[CDC_resistance$specnum==.x,a]=='R')>0)
  single <- 1*(colSums(single_colony_cdc[single_colony_cdc$Sample==.y,a]=='R')>0)
  tibble(
    type=rep(c('single','deep'), each=length(a)),
    specnum=rep(.x, 2*length(a)),
    antibiotic=c(names(single), names(deep)),
    res=c(single, deep)
  )
}) %>% group_by(type, antibiotic) %>%
  summarise(
    res_count=sum(res),
    total_count=n()
  )

temp_res_single_deep_counts$antibiotic <- gsub("_SIR", "", gsub("WGS_", "", temp_res_single_deep_counts$antibiotic))

diff_order <- temp_res_single_deep_counts %>% 
  group_by(antibiotic) %>%
  summarise(
    diff_prop=abs(diff(res_count))/sum(res_count[type=='single'])
  ) %>%
  arrange(-diff_prop)
temp_res_single_deep_counts$antibiotic <- factor(temp_res_single_deep_counts$antibiotic, levels = diff_order$antibiotic)

single_sum <- sum(temp_res_single_deep_counts$res_count[temp_res_single_deep_counts$type=='single'])
deep_sum <- sum(temp_res_single_deep_counts$res_count[temp_res_single_deep_counts$type=='deep'])
deep_sum-single_sum
(deep_sum-single_sum)/single_sum


temp_res_single_deep_counts$type[temp_res_single_deep_counts$type=='single'] <- 'single colony'
temp_res_single_deep_counts$type[temp_res_single_deep_counts$type=='deep'] <- 'plate sweep'

ggplot(temp_res_single_deep_counts, aes(x=antibiotic, y=res_count, fill=type)) + 
  geom_col(position = 'dodge') +
  theme_clean(base_size = 16) +
  theme(plot.background = element_blank(),
        legend.background = element_blank()) +
  scale_fill_manual(values=c('#ca0020', '#0571b0')) +
  scale_x_discrete(guide = guide_axis(n.dodge = 2)) +
  xlab('antibiotic category') +
  ylab("number of samples with resistance determinants")

ggsave("./figures/res_counts_deep_vs_single_MOI_1.png", width = 12, height = 7)
ggsave("./figures/res_counts_deep_vs_single_MOI_1.pdf", width = 12, height = 7)
```

We can also test if there is a significant association with the method used and the number of resistance calls in the set of single lineage samples

```{r}
summary(glm(res_count ~ type + antibiotic, temp_res_single_deep_counts, family = poisson))
```

## Association between Resistance and Multiple Carriage

We can test whether resistance is associated with multiple GPSC's being present for each resistance class. We control for carriage using a mixed model with the host as a random effect. We use logistic regression with a random effect to control for host effects and a loglog binomial link function as described in the lineage deconvolution Rmarkdown to account for multiplicity of infection.

```{r}
spec_CDC_resistance <- CDC_resistance %>%
  group_by(specnum) %>%
  group_map(~ {
    map_dfr(a, function(anti){
      df <- tibble(
        antibiotic = anti,
        num_gpscs = nrow(.x),
        gpscs = list(.x$GPSC),
        calls = list(unique(unlist(.x[,anti])))
      )
      return(df)
    }) %>% 
      add_column(host=unique(.x$host), .before = 1) %>% 
      add_column(specnum=unique(.x$specnum), .before = 1)
  }, keep=TRUE)

spec_CDC_resistance <- do.call(rbind, spec_CDC_resistance)
spec_CDC_resistance$has_res <- map_lgl(spec_CDC_resistance$calls, ~ {'R' %in% .x})
spec_CDC_resistance$treated <- merged_serotype_calls$treatment[match(spec_CDC_resistance$specnum, merged_serotype_calls$specnum)]=="TREATED"
```

We also control for population structure using by taking the first 10 dimensions after running Multidimensional Scaling (MDS) on a pairwise distance matrix inferred using Mash (v2.2.0). 

```
for r1 in *_1.fastq.gz
do
prefix=$(basename $f _1.fastq.gz)
r2="${r1/_1.fastq.gz/_2.fastq.gz}"
mash sketch -r -o "./mash_sketches/${prefix}.msh" $r1 $r2
done

mash triangle -E -p 60 ./mash_sketches/*.msh > pairwise_mash_dist.tab
```

```{r, eval=FALSE}
library(data.table)
dist_df <- fread("pairwise_mash_dist.tab", data.table = FALSE)

dist_df$V1 <- gsub("\\.fastq\\.gz", "", gsub(".*/","",dist_df$V1))
dist_df$V2 <- gsub("\\.fastq\\.gz", "", gsub(".*/","",dist_df$V2))
mdist <- max(dist_df$V3)
samples <- unique(c(dist_df$V1, dist_df$V2))
d <- matrix(mdist, nrow=length(samples), ncol = length(samples))
rownames(d) <- colnames(d) <- samples
d[as.matrix(dist_df[,c(1,2)])] <- dist_df$V3
d[as.matrix(dist_df[,c(2,1)])] <- dist_df$V3

mds <- cmdscale(d, k=10)
rownames(mds) <- gsub("_1$", "", rownames(mds))

mds_df <- as.data.frame(cbind(data.frame(sample=rownames(mds)), as.data.frame(mds)))

fwrite(mds_df, file = "./data/pairwise_mash_mds.csv", sep=',', quote = FALSE)
```

```{r}
mash_mds <- fread("./data/pairwise_mash_mds.csv") %>% as_tibble()
mash_mds$specnum_temp <- CDC_resistance$specnum[match(mash_mds$sample, CDC_resistance$lane)]
temp <- mash_mds[match(spec_CDC_resistance$specnum, mash_mds$specnum_temp),]
spec_CDC_resistance <- cbind(spec_CDC_resistance, temp)
spec_CDC_resistance$specnum_temp <- NULL


form <- as.formula(paste(c('has_res ~ multi + treated', sprintf('V%d', 1:10), '(1|host)' , 'offset(log(num_gpscs))'), collapse = " + "))

multi_res_test <- map_dfr(unique(spec_CDC_resistance$antibiotic), ~{
  print(.x)
  temp_data <- spec_CDC_resistance %>% filter(antibiotic==.x)
  temp_data$multi <- temp_data$num_gpscs>1
  temp_data$not_res <- !temp_data$has_res*1
  
  m <- tryCatch({
    model <- glmer(form, data = temp_data, family = binomial(link="cloglog"), nAGQ = 0)
    conv <- paste(model@optinfo$conv$lme4$messages, collapse = ' / ')
    model <- tidy(model) %>%
      filter(term=='multiTRUE') %>%
      add_column(antibiotic=.x, .before = 1) %>%
      add_column(converged=conv)
  },
  error = function(e) {print(paste("glm error! ", e));
    tibble()
  })
  return(m)
}) %>% 
  arrange(p.value) %>%
  mutate(adj.p=p.adjust(p.value, method="BH"))

multi_res_test$estimate <- format(multi_res_test$estimate, digits = 1)
multi_res_test$p.value <- format(multi_res_test$p.value, digits = 3)
multi_res_test$adj.p <- format(multi_res_test$adj.p, digits = 3)
multi_res_test$group <- NULL

multi_res_test$antibiotic <- gsub("_SIR", "", gsub("WGS_", "", multi_res_test$antibiotic))

knitr::kable(multi_res_test, format.args = list(decimal.mark = ".", big.mark = "'"))
write.csv(multi_res_test, file="data/multi_CDC_res_test_glmer.csv", row.names = FALSE, quote = FALSE)
```

While the assignment of resistance calls to lineages within mixtures is not perfect they are accurate enough to be worth plotting in order to verify the results of the model used above. We can plot the frequency of resistant and susceptible lineages separated out into the different antibiotic categories. The resulting box-plots indicate that resistant lineages are more often found at 100% frequency which is consistent with resistance lineages being found less often in multiple carriage than we would expect given their background frequency in the population. 

```{r}
summary_CDC <- CDC_resistance[,c(a,'lane', 'GPSC', 'specnum', 'host', 'consensus_percentage')]

plotdf <- melt(summary_CDC, id.vars=c('specnum','GPSC','lane','consensus_percentage', 'host')) %>%
  filter(value %in% c('S','R'))

plotdf$variable <- gsub("WGS_", "", plotdf$variable)
plotdf$variable <- gsub("_SIR", "", plotdf$variable)

plotdf$call <- ifelse(plotdf$value=='R', 'resistant', 'susceptible')

plotdf <- plotdf %>% filter(variable %in% multi_res_test$antibiotic[as.numeric(multi_res_test$adj.p)<0.05])


ggplot(plotdf, aes(x=call, y=consensus_percentage, colour=call)) +
  ggbeeswarm::geom_quasirandom(alpha=0.2, varwidth = TRUE, nbins = 100) +
  geom_boxplot(col='black', outlier.colour = NA, alpha=0.1, size=0.75, varwidth = TRUE) +
  facet_wrap(~variable, scales = "free_x", ncol = 2) +
  theme_clean(base_size = 16) +
  theme(plot.background = element_blank(),
        legend.background = element_blank()) +
  scale_colour_manual(values = c('#ca0020', '#0571b0')) +
  xlab("") +
  ylab("percentage") 

ggsave(filename = "./figures/CDC_resistance_box.pdf", device = "pdf", width = 7, height = 15)
ggsave(filename = "./figures/CDC_resistance_box.png", device = "png", width = 7, height = 15)


pdf2 <- plotdf %>% group_by(call, variable) %>%
  summarise(
    frac=sum(consensus_percentage>99, na.rm = TRUE)/n()
  )

ggplot(pdf2, aes(x=variable, y=frac, col=call)) +
  geom_point(size=5) +
  scale_colour_manual(values = c('#ca0020', '#0571b0')) +
  theme_clean(base_size = 20) +
  theme(plot.background = element_blank(),
        legend.background = element_blank()) +
  xlab("") +
  ylab("fraction of single lineage carriage")


ggsave(filename = "./figures/CDC_resistance_dotplot.pdf", device = "pdf", width = 10, height = 7)
ggsave(filename = "./figures/CDC_resistance_dotplot.png", device = "png", width = 10, height = 7)
```


## Using Ariba to Investigate the Association between Resistance and Multiple Carriage

We resistance calls obtained using Ariba in a similar way.

```{r}
ariba_results <- fread("./data/combined_ariba_reports.tsv.gz", data.table = FALSE) %>% as_tibble()
swab_summary$host <- paste(swab_summary$codenum, swab_summary$category, sep="_")
ariba_results$specnum <- manifests_qc$`DONOR ID (required for EGA)`[match(ariba_results$sample, manifests_qc$`Lane Name`)]
ariba_results$host <- swab_summary$host[match(ariba_results$specnum, swab_summary$specnum)]
ariba_results$serotype <- map(ariba_results$specnum, ~ unique(merged_serotype_calls$serotype[(merged_serotype_calls$specnum==.x) & !is.na(merged_serotype_calls$serotype)]))
ariba_results$GPSCs <- map(ariba_results$specnum, ~ unique(merged_serotype_calls$GPSC[(merged_serotype_calls$specnum==.x) & !is.na(merged_serotype_calls$GPSC)]))

ariba_results$ngpscs <- map_int(ariba_results$GPSCs, length)

ariba_results_filt <- ariba_results %>% filter(((var_only==1) & (known_var==1)) | (var_only==0))

temp_pa <- spec_CDC_resistance[!duplicated(spec_CDC_resistance$specnum),]
temp_pa$antibiotic <- 'ery'
temp_pa$has_res <- temp_pa$specnum %in% (ariba_results_filt %>% filter(grepl(".*mefA.*", ariba_ref_name) |
                                       grepl(".*ermB.*", ariba_ref_name)))$specnum
ariba_res_pa <- temp_pa

temp_pa$antibiotic <- 'tet'
temp_pa$has_res <- temp_pa$specnum %in% (ariba_results_filt %>% filter(grepl(".*tet.*", ariba_ref_name)))$specnum
ariba_res_pa <- rbind(ariba_res_pa, temp_pa)

temp_pa$antibiotic <- 'cot'
temp_pa$has_res <- temp_pa$specnum %in% (ariba_results_filt %>% filter(grepl(".*fol.*", ariba_ref_name)))$specnum
ariba_res_pa <- rbind(ariba_res_pa, temp_pa)

temp_pa$antibiotic <- 'rif'
temp_pa$has_res <- temp_pa$specnum %in% (ariba_results_filt %>% filter(grepl(".*rpoB.*", ariba_ref_name)))$specnum
ariba_res_pa <- rbind(ariba_res_pa, temp_pa)

temp_pa$antibiotic <- 'lfx'
temp_pa$has_res <- temp_pa$specnum %in% (ariba_results_filt %>% filter(grepl(".*gyr.*", ariba_ref_name) |
                                             grepl(".*par.*", ariba_ref_name)))$specnum
ariba_res_pa <- rbind(ariba_res_pa, temp_pa)

form <- as.formula(paste(c('not_res ~ multi + treated', sprintf('V%d', 1:10), '(1|host)'), collapse = " + "))


multi_ariba_res_test <- map_dfr(unique(ariba_res_pa$antibiotic), ~{
  temp_data <- ariba_res_pa %>% filter(antibiotic==.x)
  temp_data$multi <- temp_data$num_gpscs>1
  temp_data$not_res <- !temp_data$has_res
  
  if(nrow(table(temp_data$has_res, temp_data$multi))<2) return(tibble())
  
  tidy(glmer(form, data = temp_data, family = binomial(link = logexp(temp_data$num_gpscs)), nAGQ = 0)) %>%
    filter(term=='multiTRUE') %>%
    add_column(antibiotic=.x ,.before = 1)
}) %>% 
  arrange(p.value) %>%
  mutate(adj.p=p.adjust(p.value, method="BH"))

multi_ariba_res_test$estimate <- format(multi_ariba_res_test$estimate, digits = 1)
multi_ariba_res_test$p.value <- format(multi_ariba_res_test$p.value, digits = 3)
multi_ariba_res_test$adj.p <- format(multi_ariba_res_test$adj.p, digits = 3)
multi_ariba_res_test$group <- NULL

knitr::kable(multi_ariba_res_test, format.args = list(decimal.mark = ".", big.mark = "'"))
write.csv(multi_ariba_res_test, file="data/multi_ariba_res_test_glmer.csv", row.names = FALSE, quote = FALSE)
```

## Resistance in Mothers and Children

We can also look at the frequency of resistance in mothers as compared to their children.

```{r}
CDC_resistance$mother_child <- map_chr(str_split(CDC_resistance$host, '-'), ~.x[[length(.x)]])

resistance_by_host_gpsc <- do.call(rbind, CDC_resistance %>% 
  group_by(host, GPSC, mother_child) %>%
  group_map(~{
    as_tibble(t(colSums(.x[,a]=='R')>0)) %>%
      add_column(mother_child=unique(.x$mother_child), .before=1) %>%
      add_column(GPSC=unique(.x$GPSC), .before=1) %>%
      add_column(host=unique(.x$host), .before=1)
  }, keep = TRUE))


plotdf <- melt(resistance_by_host_gpsc, id.vars=c("host", "GPSC", "mother_child"), variable.name="antibiotic", value.name="resistant")
plotdf$antibiotic <- gsub("_SIR", "", gsub("WGS_", "", plotdf$antibiotic))

plotdf <- plotdf %>% group_by(host, mother_child, antibiotic) %>%
  summarise(resistant=any(resistant),
            ngpscs=n(), .groups = 'keep')

mother_child_test <- map_dfr(unique(plotdf$antibiotic), function(anti){
  testdf <- plotdf %>% filter(antibiotic==anti)
  broom::tidy(glm(resistant ~ mother_child + ngpscs, data = testdf, family = 'binomial')) %>% 
    filter(term=='mother_childMother') %>%
    add_column(antibiotic=anti, .before = 1)
}) %>% arrange(p.value)

mother_child_test$p.adj <- p.adjust(mother_child_test$p.value, method = "BH")
write.csv(mother_child_test, file="data/mother_vs_child_resistance_glm.csv", quote = FALSE, row.names = FALSE)

ggplot(plotdf, aes(x=resistant, fill=resistant)) +
  geom_bar(position = position_dodge(preserve = "single")) +
  facet_grid(antibiotic~mother_child, scales = "free") +
  theme_clean(base_size = 16) +
  theme(plot.background = element_blank(),
        legend.background = element_blank()) +
  xlab("") +
  scale_fill_manual(values=c('#0571b0', '#ca0020')) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        strip.text.y = element_text(angle=0))

ggsave("./figures/mother_vs_child_resistance.pdf", width = 7, height = 15)
ggsave("./figures/mother_vs_child_resistance.png", width = 7, height = 15)
```

## Virulence

We can also look at the association between resistance an virulence elements identified using Ariba

```{r}
virulence_ariba <- fread("./data/combined_virulence_ariba_reports.tsv.gz") %>% as_tibble() %>% 
  filter(((var_only==1) & (known_var==1)) | (var_only==0))
virulence_ariba <- virulence_ariba[grepl("Streptococcus_pneumoniae.*", virulence_ariba$ariba_ref_name),]

virulence_ariba_summary <- virulence_ariba %>% 
  group_by(sample, cluster) %>%
  summarise(
    n_var_calls = n(),
    var_description = list(unique(var_description)),
    var_description = list(unique(var_description)),
    ref_ctg_change = list(unique(ref_ctg_change)),
    ref_ctg_effect = list(ref_ctg_effect),
    ref_start = list(unique(ref_start)),
    ctg_nt = list(unique(ctg_nt)),
    free_text = list(unique(free_text))
  )

virulence_ariba_summary$specnum <- manifests_qc$`DONOR ID (required for EGA)`[match(virulence_ariba_summary$sample,manifests_qc$`Lane Name`)]
virulence_ariba_summary$host <- swab_summary$ID[match(virulence_ariba_summary$specnum,swab_summary$specnum)]
virulence_ariba_summary$treated <- swab_summary$treatment[match(virulence_ariba_summary$specnum,swab_summary$specnum)]=='TREATED'

vir_clusters <- unique(virulence_ariba_summary$cluster)
specnums <- unique(virulence_ariba_summary$specnum)
virulence_ariba_pa <- map_dfr(vir_clusters, ~{
  tibble(
    virulence = .x,
    specnum = specnums,
    has_vir = specnums %in% virulence_ariba_summary$specnum[virulence_ariba_summary$cluster==.x]
  )
})
virulence_ariba_pa$host <- swab_summary$ID[match(virulence_ariba_pa$specnum,swab_summary$specnum)]
virulence_ariba_pa$treated <- swab_summary$treatment[match(virulence_ariba_pa$specnum,swab_summary$specnum)]=='TREATED'
virulence_ariba_pa$nsero <- map_dbl(virulence_ariba_pa$specnum, ~ sum(merged_serotype_calls$specnum==.x & !is.na(merged_serotype_calls$consensus_percentage)))
virulence_ariba_pa <- virulence_ariba_pa %>% filter(nsero>0)
temp_merged <- merged_serotype_calls %>% arrange(-consensus_percentage)
virulence_ariba_pa$gpsc <- map(virulence_ariba_pa$specnum, ~ temp_merged$GPSC[temp_merged$specnum==.x])

gs <- as.character(unique(unlist(virulence_ariba_pa$gpsc)))
gs <- gs[!is.na(gs)]
ss <- unique(unlist(virulence_ariba_pa$specnum))
x <- matrix(FALSE, nrow = length(ss), ncol = length(gs), dimnames = list(ss, gs))
x[cbind(rep(virulence_ariba_pa$specnum, map_dbl(virulence_ariba_pa$gpsc, length)), as.character(unlist(virulence_ariba_pa$gpsc)))] <- TRUE
pc <- prcomp(x)
pc <- as_tibble(pc$x[,1:10])
pc$specnum <- ss
virulence_ariba_pa <- merge(virulence_ariba_pa, pc, by.x="specnum", by.y="specnum")

mash_mds <- fread("./data/pairwise_mash_mds.csv") %>% as_tibble()
mash_mds$specnum_temp <- manifests_qc$`DONOR ID (required for EGA)`[match(mash_mds$sample, manifests_qc$`Lane Name`)]
temp <- mash_mds[match(virulence_ariba_pa$specnum, mash_mds$specnum_temp),]
virulence_ariba_pa <- cbind(virulence_ariba_pa, temp)
virulence_ariba_pa$specnum_temp <- NULL

form <- as.formula(paste(c('not_virulent ~ multi + treated', paste('V', 1:10, sep=""), '(1|host)'), collapse = " + "))
  
ariba_vir_tests <- map_dfr(vir_clusters, ~{
  temp_data <- virulence_ariba_pa %>% filter(virulence==.x)
  temp_data$not_virulent <- !temp_data$has_vir
  temp_data$multi <-temp_data$nsero>1 
  
  tryCatch({
    tidy(glmer(form, data = temp_data, family = binomial(link = logexp(temp_data$nsero)), nAGQ = 0)) %>%
      filter(term=='multiTRUE') %>%
      add_column(virulent=.x, .before = 1)
  },
  warning = function(x) {tibble(virulent=.x, term='multiTRUE', estimate=NA, std.error=NA, statistic=NA, p.value=NA, group='fixed')},
  error = function(x) {tibble(virulent=.x, term='multiTRUE', estimate=NA, std.error=NA, statistic=NA, p.value=NA, group='fixed')})
}) %>% 
  arrange(p.value) %>%
  mutate(adj.p=p.adjust(p.value, method="BH"))

tbl <- ariba_vir_tests %>% filter(adj.p<0.05)
tbl$p.value <- format(tbl$p.value, digits=2)
tbl$adj.p <- format(tbl$adj.p, digits=2)
knitr::kable(tbl)

write.csv(ariba_vir_tests, file = "./data/ariba_vir_glmer_tests_mash_mds.csv", quote = FALSE, row.names = FALSE)
```